### PR TITLE
Tear down connections

### PR DESCRIPTION
**Pull request by llerrowa**
This PR addresses a memory leak in the duckdb connection manager (this is more prevalent when using an in memory database). 

Whenever a connection is created a reference to it is stored by the `ConnectionManager` in the `connections` vector. When a connection is closed the `ConnectionManager` still keeps a reference to the connection and this is not cleared*, this PR adds a deconstruction to remove the entry in the `connections` vector.

The memory leak can be seen by opening an in memory database and connecting / disconnecting to this many times, you will see the memory of your application increase and it will not be freed up. 

* calling `GetConnectionList()` will erase any disposed pointers but this is called `Checkpoint` which doesn't run for in memory database via `LockClients`.

### CODE REVIEW

File: src/main/connection.cpp [hunk 1]
```diff
@@ -26,6 +26,10 @@ Connection::Connection(DatabaseInstance &database) : context(make_shared<ClientC
 Connection::Connection(DuckDB &database) : Connection(*database.instance) {
 }
 
+Connection::~Connection() {
+	ConnectionManager::Get(*database).RemoveConnection(*context);
+}
+
 string Connection::GetProfilingInformation(ProfilerPrintFormat format) {
```
Comment (Mytherin) [inline src/main/connection.cpp:33]: Could you add a test case where we open and close a bunch of new connections many times?

**Review by Mytherin — COMMENTED** [review]
Thanks for the PR! Looks good to me. Some comments:

File: src/main/connection.cpp (hunk 1 above)
Comment (llerrowa) [reply src/main/connection.cpp:33]: I've created a test to cover this, let me know what you think

**Review by Mytherin — APPROVED** [review]
Looks good to me. Can be merged if the tests pass.

**Pull request merged**